### PR TITLE
Bybit funding rate data source implementation

### PR DESCRIPTION
**Pull request by arodus**
This implementation is based on the binance funding rate source. 

**Related Issue**
https://github.com/QuantConnect/Lean/issues/5898

### CODE REVIEW

File: DataProcessing/BybitFundingRateDownloader.cs [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+    /// <summary>
+    /// Runs the instance of the object.
+    /// </summary>
+    /// <returns>True if process all downloads successfully</returns>
+    public bool Run()
+    {
+        var ratePerSymbol = new Dictionary<string, Dictionary<DateTime, decimal>>();
+        foreach (var date in GetProcessingDates())
+        {
+            foreach (var apiFundingRate in GetData(date))
```
Comment (arodus) [inline DataProcessing/BybitFundingRateDownloader.cs]: The process can be sped up immensely by fetching the data for multiple days. I can add this if it's desired.

Comment (Martin-Molinero) [reply DataProcessing/BybitFundingRateDownloader.cs]: I think as is it's okay 👍  the data jobs will run with the single `_deploymentDate` set

File: DataProcessing/BybitFundingRateDownloader.cs [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+
+        var tempPath = Path.Combine(Path.GetTempPath(), $"{Guid.NewGuid()}.tmp");
+        File.WriteAllLines(tempPath, finalLines);
+        var tempFilePath = new FileInfo(tempPath);
+        tempFilePath.MoveTo(finalPath, true);
+    }
+
+    private IEnumerable<BybitInstrumentInfo> GetExchangeInfo()
+    {
+        var url = $"{BybitApiEndpoint}/v5/market/instruments-info?limit=1000&status=Trading";
```
Comment (Martin-Molinero) [inline DataProcessing/BybitFundingRateDownloader.cs]: > limit=1000

Think we should handle this limit being hit, else we would silently start skipping symbols when the limit is hit

File: DataProcessing/BybitFundingRateDownloader.cs [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+    private IEnumerable<DateTime> GetProcessingDates()
+    {
+        if (_deploymentDate.HasValue)
+        {
+            return new[] { _deploymentDate.Value };
+        }
+        else
+        {
+            // everything
+            return Time.EachDay(new DateTime(2019, 9, 13), DateTime.UtcNow.Date);
```
Comment (Martin-Molinero) [inline DataProcessing/BybitFundingRateDownloader.cs]: The start date should match when the bybit exchange started having this data available, trading the perpetual futures, tested for example `2019, 1, 1` and it seems the API is returning data too?

File: DataProcessing/BybitFundingRateDownloader.cs [hunk 4]
```diff
@@ -0,0 +1,333 @@
+﻿using System;
```
Comment (Martin-Molinero) [inline DataProcessing/BybitFundingRateDownloader.cs]: Minor missing the license header 👍

**Review by Martin-Molinero — COMMENTED** [review]
Nice! Minor comments 👍

File: DataProcessing/BybitFundingRateDownloader.cs (hunk 2 above)
Comment (arodus) [reply DataProcessing/BybitFundingRateDownloader.cs]: This is very unlikely to happen when we're fetching the data for each day separately. All the pairs currently have a funding interval of 8 hours meaning we have 3 entries per response. And I don't think they add a coin with a funding interval of 1 minute.

Comment (Martin-Molinero) [reply DataProcessing/BybitFundingRateDownloader.cs]: This call & endpoint is to get the instruments to fetch the funding rate for though, not the funding rate itself I mean

File: DataProcessing/BybitFundingRateDownloader.cs (hunk 3 above)
Comment (arodus) [reply DataProcessing/BybitFundingRateDownloader.cs]: The API is not returning data but actually returning with an error that there is no data available which is just being ignored. The question would also be there also how the downloader should handle hard errors. Just throw? Right now pretty much everything besides network errors is not being handled. 

I also double-checked when the first data is available, it's `2019,11,14`  will update it.

**Review by Martin-Molinero — APPROVED** [review]
Thank you! 🚀

**Pull request merged**